week06_policy_based/reinforce_tensorflow.ipynb

```diff
@@ -1,592 +1,387 @@
+    "    compute cumulative rewards R(s,a) (a.k.a. G(s,a) in Sutton '16)\n",
+    "    R_t = r_t + gamma*r_{t+1} + gamma^2*r_{t+2} + ...\n",
+    "\n",
+    "    The simple way to compute cumulative rewards is to iterate from last to first time tick\n",
+    "    and compute R_t = r_t + gamma*R_{t+1} recurrently\n",
+    "\n",
+    "    You must return an array/list of cumulative rewards with as many elements as in the initial rewards.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    <your code here >\n",
+    "\n",
+    "    return < array of cumulative rewards >"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "assert len(get_cumulative_rewards(range(100))) == 100\n",
+    "assert np.allclose(get_cumulative_rewards([0, 0, 1, 0, 0, 1, 0], gamma=0.9), [\n",
+    "                   1.40049, 1.5561, 1.729, 0.81, 0.9, 1.0, 0.0])\n",
+    "assert np.allclose(get_cumulative_rewards(\n",
+    "    [0, 0, 1, -2, 3, -4, 0], gamma=0.5), [0.0625, 0.125, 0.25, -1.5, 1.0, -4.0, 0.0])\n",
+    "assert np.allclose(get_cumulative_rewards(\n",
+    "    [0, 0, 1, 2, 3, 4, 0], gamma=0), [0, 0, 1, 2, 3, 4, 0])\n",
+    "print(\"looks good!\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def train_step(_states, _actions, _rewards):\n",
+    "    \"\"\"given full session, trains agent with policy gradient\"\"\"\n",
+    "    _cumulative_rewards = get_cumulative_rewards(_rewards)\n",
+    "    update.run({states: _states, actions: _actions,\n",
+    "                cumulative_rewards: _cumulative_rewards})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Playing the game"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_session(t_max=1000):\n",
+    "    \"\"\"play env with REINFORCE agent and train at the session end\"\"\"\n",
+    "\n",
+    "    # arrays to record session\n",
+    "    states, actions, rewards = [], [], []\n",
+    "\n",
+    "    s = env.reset()\n",
+    "\n",
+    "    for t in range(t_max):\n",
+    "\n",
+    "        # action probabilities array aka pi(a|s)\n",
+    "        action_probas = get_action_proba(s)\n",
+    "\n",
+    "        a = <pick random action using action_probas >\n",
+    "\n",
+    "        new_s, r, done, info = env.step(a)\n",
+    "\n",
+    "        # record session history to train later\n",
+    "        states.append(s)\n",
+    "        actions.append(a)\n",
+    "        rewards.append(r)\n",
+    "\n",
+    "        s = new_s\n",
+    "        if done:\n",
+    "            break\n",
+    "\n",
+    "    train_step(states, actions, rewards)\n",
+    "\n",
+    "    return sum(rewards)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
```

Comment:
The only output that may be significant but was removed is this one. Personally, I think it is not super important, but @justheuristic's opinion would be nice to hear.

Reply:
i concur, this one is not significant